File: ios/App/App/DeviceStoragePlugin.swift
import Foundation
import Capacitor

/// Reports total/available capacity of the volume N-Go's app data lives on, so storageHealth.ts
/// can use real device free space instead of budget-only mode. See
/// src/services/nativePlugins/deviceStorage.ts for the JS side and
/// docs/OFFLINE_STORAGE_MANAGEMENT.md for the design.
///
/// Measures the volume containing the app's documents directory. Every directory inside an iOS
/// app container (Documents, Library, tmp — so both Capacitor Filesystem's Directory.Data and the
/// WebView's IndexedDB) sits on the same data volume, so one lookup covers all of N-Go's writes.
///
/// AVAILABLE capacity uses `volumeAvailableCapacityForImportantUsage` (iOS 11+), not raw
/// `volumeAvailableCapacity`. That is deliberate and is the API Apple documents for exactly this
/// question — "can I store something the user asked for?" — because it accounts for space the
/// system can reclaim by purging caches/offloadable content. It is therefore usually LARGER than
/// raw free bytes and is a better operational estimate of what N-Go can still write than a raw
/// figure that would report a device as full while iOS is still holding purgeable data. If that
/// key is unavailable the code falls back to plain `volumeAvailableCapacity`, then rejects.
///
/// Privacy: returns two integers. No iCloud query, no file enumeration, no identifiers, and no
/// permission prompt — volume capacity for the app's own container requires no entitlement.
@objc(DeviceStoragePlugin)
public class DeviceStoragePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "DeviceStoragePlugin"
    public let jsName = "DeviceStorage"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "getStorageInfo", returnType: CAPPluginReturnPromise)
    ]

    @objc func getStorageInfo(_ call: CAPPluginCall) {
        guard let containerURL = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first else {
            call.reject("Application documents directory is unavailable")
            return
        }

        do {
            let values = try containerURL.resourceValues(forKeys: [
                .volumeTotalCapacityKey,
                .volumeAvailableCapacityForImportantUsageKey,
                .volumeAvailableCapacityKey
            ])

            guard let totalCapacity = values.volumeTotalCapacity, totalCapacity > 0 else {
                call.reject("Volume reported no usable total capacity")
                return
            }

            // Prefer "important usage" (Int64); fall back to plain available capacity (Int).
            let availableBytes: Int64
            if let importantUsage = values.volumeAvailableCapacityForImportantUsage {
                availableBytes = importantUsage
            } else if let plainAvailable = values.volumeAvailableCapacity {
                availableBytes = Int64(plainAvailable)
            } else {
                call.reject("Volume reported no available-capacity value")
                return
            }

            call.resolve([
                "totalBytes": Int64(totalCapacity),
                "freeBytes": max(0, availableBytes)
            ])
        } catch {
            // The JS side treats a rejection as "unavailable" and falls back to budget-only health.
            call.reject("Failed to read device storage information: \(error.localizedDescription)")
        }
    }
}

File: android/app/src/main/java/com/strata/ngo/field/dev/DeviceStoragePlugin.java
package com.strata.ngo.field.dev;

import android.os.StatFs;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.io.File;

/**
 * Reports total/available capacity of the filesystem N-Go's app data actually lives on, so
 * storageHealth.ts can use real device free space instead of budget-only mode. See
 * src/services/nativePlugins/deviceStorage.ts for the JS side and
 * docs/OFFLINE_STORAGE_MANAGEMENT.md for the design.
 *
 * Measures getFilesDir() — the app's INTERNAL data volume, which is where Capacitor Filesystem's
 * Directory.Data (mediaStore.ts's captured media, configMediaCache.ts's reference media) and the
 * WebView's IndexedDB both land. Deliberately NOT getExternalFilesDir()/removable storage: an SD
 * card's free space says nothing about whether N-Go can still write.
 *
 * Uses StatFs#getTotalBytes()/#getAvailableBytes() (API 18+, well under this project's minSdk 24)
 * rather than the deprecated getBlockCount()*getBlockSize() multiplication, which additionally
 * overflows int on large volumes.
 *
 * Permissions: NONE. Reading stat info for the app's own private data directory requires no
 * runtime permission and no manifest entry on any supported API level — no MANAGE_EXTERNAL_STORAGE,
 * no READ/WRITE_EXTERNAL_STORAGE. The plugin never enumerates files; it returns two longs.
 */
@CapacitorPlugin(name = "DeviceStorage")
public class DeviceStoragePlugin extends Plugin {

    @PluginMethod
    public void getStorageInfo(PluginCall call) {
        try {
            File dataDir = getContext().getFilesDir();
            if (dataDir == null) {
                call.reject("Application files directory is unavailable");
                return;
            }

            StatFs stat = new StatFs(dataDir.getAbsolutePath());
            long totalBytes = stat.getTotalBytes();
            long availableBytes = stat.getAvailableBytes();

            if (totalBytes <= 0) {
                call.reject("Filesystem reported a non-positive total capacity");
                return;
            }

            JSObject result = new JSObject();
            // JSObject.put(String, long) is preserved as a JSON number; the JS side re-validates
            // both figures anyway (isValidNativeStorageInfo in deviceStorageCapability.ts).
            result.put("totalBytes", totalBytes);
            result.put("freeBytes", Math.max(0L, availableBytes));
            call.resolve(result);
        } catch (Exception e) {
            // IllegalArgumentException (path vanished), SecurityException, or anything else — the
            // JS side treats a rejection as "unavailable" and falls back to budget-only health.
            call.reject("Failed to read device storage information: " + e.getMessage(), e);
        }
    }
}
